File: patterns/03-data-messaging-patterns/01-transactional-outbox/go/transactional_outbox.go
package transactionaloutbox

type PatternError struct{ code string }

func (e *PatternError) Error() string { return e.code }
func (e *PatternError) Code() string  { return e.code }

func Evaluate(input map[string]any) (any, error) {
	raw, _ := input["relay_attempts"].([]any)
	if input["commit"] != true {
		if len(raw) > 0 {
			return nil, &PatternError{code: "relay_without_commit"}
		}
		return map[string]any{"aggregate_state": "absent", "outbox_status": "absent", "deliveries": []string{}, "relay_receipts": []string{}}, nil
	}
	messageID := input["message_id"].(string)
	status := "pending"
	deliveries := []string{}
	receipts := []string{}
	for _, item := range raw {
		attempt := item.(string)
		if status == "sent" {
			receipts = append(receipts, "skipped_already_sent:"+messageID)
			continue
		}
		switch attempt {
		case "crash_before_publish":
			receipts = append(receipts, "crash_before_publish:"+messageID)
		case "crash_after_publish":
			deliveries = append(deliveries, messageID)
			receipts = append(receipts, "crash_after_publish:"+messageID)
		case "success":
			deliveries = append(deliveries, messageID)
			receipts = append(receipts, "published:"+messageID)
			status = "sent"
		default:
			return nil, &PatternError{code: "unknown_relay_outcome"}
		}
	}
	return map[string]any{"aggregate_state": input["new_state"], "outbox_status": status, "deliveries": deliveries, "relay_receipts": receipts}, nil
}

File: patterns/03-data-messaging-patterns/01-transactional-outbox/js/transactional_outbox.mjs
export class PatternError extends Error { constructor(code) { super(code); this.code = code; } }

export function evaluate(input) {
    const attempts = input.relay_attempts ?? [];
    if (input.commit !== true) {
        if (attempts.length > 0) throw new PatternError("relay_without_commit");
        return {aggregate_state: "absent", outbox_status: "absent", deliveries: [], relay_receipts: []};
    }
    const messageId = input.message_id;
    let status = "pending";
    const deliveries = [];
    const receipts = [];
    for (const attempt of attempts) {
        if (status === "sent") { receipts.push(`skipped_already_sent:${messageId}`); continue; }
        if (attempt === "crash_before_publish") receipts.push(`crash_before_publish:${messageId}`);
        else if (attempt === "crash_after_publish") { deliveries.push(messageId); receipts.push(`crash_after_publish:${messageId}`); }
        else if (attempt === "success") { deliveries.push(messageId); receipts.push(`published:${messageId}`); status = "sent"; }
        else throw new PatternError("unknown_relay_outcome");
    }
    return {aggregate_state: input.new_state, outbox_status: status, deliveries, relay_receipts: receipts};
}
